text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,171 @@
+  point, exit gracefully
+* upgrade activation points should be defined in terms of era IDs, not block heights
+* config.toml updated to take the path to the chainspec upgrade file, similar to the `node.chainspec_config_path`
+  option.  Alternatively, just remove this option, and require the chainspec, its upgrade file, accounts.csv and the
+  config.toml file all to live in the same folder
+
+### Data migration
+
+We require to add a new `migrate-data` subcommand to `casper-node`.  We could provide a separate binary to be used
+exclusively for data and config migration, but it seems simpler to just make this a subcommand.  Simpler for the
+installer, for the `upgrade-bootstrapper` and simpler to code.
+
+Migration of any of the databases needs to take the following approach by iterating each value in the DB and doing:
+
+1. parse as the old type
+1. convert it to the new type
+1. store the new type, possibly under a different key
+1. if stored under a different key (i.e. it's not just an overwrite), then delete the old value
+
+We need to be able to resume migration in case it's killed while running.  This means we need to ensure that we either
+keep track of our progress by writing to disk after each individual item is successfully migrated, or else we ensure
+that parsing a new value as an old value fails, or else we ensure that running the conversion step on an
+already-converted item is a no-op.
+
+If data migration is going to take a prohibitively long time to execute, we should look to have the
+`upgrade-bootstrapper` perform _data_ migration only (not config migration) as soon as the installer provides the new
+`casper-node` binary, while the current `casper-node` is still running in validator mode.  However, this might result in
+a doubling or more of the disk space consumed by the databases, since the migrated data cannot be allowed to overwrite
+the existing data while the node is still running as a validator, and _this_ disk consumption might prove to be
+prohibitive.
+
+Data migration would also still need to be executed again once the old `casper-node` stops due to reaching
+the upgrade activation point, since more data will likely have been written to the databases between then and when the
+data migration was previously completed.  Running the `migrate-data` subcommand of the node could be done repeatedly at
+regular intervals to ensure that only a small amount is left to do after the old node binary has stopped.
+
+### Config migration
+
+Migration will also need to handle converting an old config file to a new one.  For this to work, we need to know which
+config values have been modified by the user from their default values.  This requires either default values to be
+provided, hard-coded in the node, or else the `migrate-data` subcommand will need to be passed not just the old and new
+config.toml files, but also a copy of each with their default values set.
```

Comment:
How about we `include_str!` the default `config.toml` somewhere in the code? This way we aren't hardcoding the defaults per se, but the binary still has the defaults contained within it.

Reply:
Interesting idea - that could well be simpler.  The only downside I can see is that I quite liked the idea of improved UX by enforcing every config value to have user-facing documentation specifying the default value.  But that's a minor point - we can still include them where they're deemed useful.